### ISSUE TITLE
proposal: 

### ISSUE THREAD

**Issue by kishaningithub** [gonum/gonum#1763]
### Background
In this package i have only used the [Binomial function](https://github.com/gonum/gonum/blob/99041dcffe021d7bf33f93de3d2f4a52f8cd69dc/stat/combin/combin.go#L29) and nothing else. 

If i import gonum using go modules i get the complete project which is quite big in my opinion

### Proposal

This is inspired by the modularization move of [aws-sdk-go-v2](https://github.com/aws/aws-sdk-go-v2) which is also a big project in a single repo (Screenshot below)
<img width="807" alt="Screenshot 2022-01-19 at 5 10 31 AM" src="https://user-images.githubusercontent.com/763760/150036244-554d87ef-b1c8-44e7-9b89-a94a0cdf3264.png">
More info - https://aws.amazon.com/tr/blogs/developer/aws-sdk-for-go-version-2-general-availability/?nc1=h_ls

### Potential impact of proposal

The people who just use a sub portion of this repo will only get those files/modules which they are actually using. This reduces the module size. For the use case i mentioned in the background, it would be great if i can just import the `combin` module and not anything else.

From an effort perspective, this would involve creation of `go.mod` files within each logical module. For the above use, this would involve creation of a `go.mod` file within the `combin` directory.

One thing i am not fully sure is on how this will affect the backwards compatibility, this is something that must be thought through.

**Comment by kortschak** [comment]
This would be a significant breaking change, it would also cause significantly more work for Gonum contributors when resolving issues since we would need to reconstruct complete states of modules from the reporter's context and also during development since there is a fair amount of cross communication between the packages within the repo.

It's worth noting that the structure of the project was originally in a set of repositories and we merged them into the small set that we have now because of the complexity involved in the previous structure.

**Comment by sbinet** [comment]
to better understand your issue, what is exactly the problem at hand?

- is it the size of the underlying `git clone` to get a couple of Go files under `stat/combin`?
- is it the size/number of the dependencies you get when `go get`-ing the Gonum module?
- something else?

AFAIU, both of the first issues will be solved with improvements of `go mod` tooling.
Eventually, the first one will be solved by pulling a zip file holding the exact revision of a given module (sans the full `git` history).
The second one will be adressed with dependencies trimming, that will ship with Go-1.18 (it's already there in some capacity with 1.17 IIRC).

As Dan alluded to, the mechanics of synchronizing `go.mod` files in repos are a bit tedious.
and rather than breaking things into smaller Go modules, I'd be sympathetic with lumping `gonum/gonum` and `gonum/plot` into a single Go module/repo. (but that's just me, I think).

**Comment by kortschak** [comment]
The alternative, if there is just one function that you are using, is to copy the file that provides that function and the LICENSE, and then you have the minimal representation of the functionality and the cost of no future automatic bug fixes.

> I'd be sympathetic with lumping `gonum/gonum` and `gonum/plot` into a single Go module/repo. (but that's just me, I think).

This will be the resolution to v1 -> v2 if that ever happens.

**Comment by sbinet** [comment]
> This will be the resolution to v1 -> v2 if that ever happens.

you mean `gonum.org/x/gonum@v0.1.0` ? :)
(besides the obvious bikeshedding bait, it would also improve our compatibility with Go modules. I vaguely remember an issue with the `/v1` in our module name at some point)

**Comment by kortschak** [comment]
Yeah, we made a decision before the landscape had settled. We were grandfathered in. What I was thinking was that if all the family end up in one repo, gonum.org/v2, then the pattern that exists now can continue.
